Testing a bug fix - searchd hangs when stopping after FREEZE and FLUSH RAMCHUNK

PavelShilin89 opened the pull request:
Added case of searhd stopping after FREEZE and FLUSH RAMCHUNK to test-unfreeze-defrosts.rec. 
Searchd correctly terminates, the bug is fixed.

sanikolaev commented on test/clt-tests/core/test-unfreeze-defrosts.rec:
```diff
@@ -62,3 +62,27 @@ sleep 5; mysql -h0 -P9306 -e "select count(*) from orders\G"
+mysql -h0 -P9306 -e "drop table if exists t; create table t; freeze t; flush ramchunk t;" &
+––– output –––
+[1] %{NUMBER}
+––– input –––
+sleep 1; searchd --stop
+––– output –––
++---------------------------------+---------------------------------+
+| file                            | normalized                      |
++---------------------------------+---------------------------------+
+| /var/lib/manticore/t/t.meta     | /var/lib/manticore/t/t.meta     |
+| /var/lib/manticore/t/t.ram      | /var/lib/manticore/t/t.ram      |
+| /var/lib/manticore/t/t.settings | /var/lib/manticore/t/t.settings |
++---------------------------------+---------------------------------+
```
This input/output pair looks wrong. I believe if you increase the sleep value to some 5 seconds the output will be different which is not good as it makes the test depend on some randomness.

sanikolaev commented on test/clt-tests/core/test-unfreeze-defrosts.rec:
```diff
@@ -62,3 +62,27 @@ sleep 5; mysql -h0 -P9306 -e "select count(*) from orders\G"
+| /var/lib/manticore/t/t.ram      | /var/lib/manticore/t/t.ram      |
+| /var/lib/manticore/t/t.settings | /var/lib/manticore/t/t.settings |
++---------------------------------+---------------------------------+
+Manticore %{SEMVER} %{COMMITDATE}#!/(\sdev)?\s/!#(columnar %{SEMVER} %{COMMITDATE}) (secondary %{SEMVER} %{COMMITDATE}) (knn %{SEMVER} %{COMMITDATE})
+Copyright (c) 2001-2016, Andrew Aksyonoff
+Copyright (c) 2008-2016, Sphinx Technologies Inc (http://sphinxsearch.com)
+Copyright (c) 2017-2024, Manticore Software LTD (https://manticoresearch.com)
+[#!/[0-9]{2}:[0-9]{2}.[0-9]{3}/!#] [#!/[0-9]+/!#] using config file '/etc/manticoresearch/manticore.conf' (%{NUMBER} chars)...
+[#!/[0-9]{2}:[0-9]{2}.[0-9]{3}/!#] [#!/[0-9]+/!#] stop: successfully sent SIGTERM to pid %{NUMBER}
```
This looks redundant. Adding `|grep TERM` to the searchd call should make the test clearer.

sanikolaev commented on test/clt-tests/core/test-unfreeze-defrosts.rec:
```diff
@@ -62,3 +62,11 @@ sleep 5; mysql -h0 -P9306 -e "select count(*) from orders\G"
 ––– output –––
 *************************** 1. row ***************************
 count(*): 10000
+––– input –––
+mysql -h0 -P9306 -e "drop table if exists t; create table t; freeze t; flush ramchunk t;" >/dev/null 2>&1 &
+––– output –––
+[1] %{NUMBER}
+––– input –––
+sleep 3; searchd --stop | grep TERM
+––– output –––
+[#!/[A-Za-z]{3}\s+[A-Za-z]{3}\s+[0-9]{1,2}\s+[0-9]{2}:[0-9]{2}:[0-9]{2}\.[0-9]{3}\s+[0-9]{4}/!#] [%{NUMBER}] stop: successfully sent SIGTERM to pid %{NUMBER}
```
As discussed, it's better to do `sleep 1; true` or smth, so you can be sure that `FREEZE` actually did something since otherwise your final test `stop: successfully sent SIGTERM to pid %{NUMBER}` can pass even if you don't do `drop table if exists t; create table t; freeze t; flush ramchunk t` at all

sanikolaev commented on test/clt-tests/core/test-unfreeze-defrosts.rec:
```diff
@@ -62,3 +62,18 @@ sleep 5; mysql -h0 -P9306 -e "select count(*) from orders\G"
+| /var/lib/manticore/t/t.settings | /var/lib/manticore/t/t.settings |
++---------------------------------+---------------------------------+
+––– input –––
+searchd --stop > /dev/null 2>&1; if timeout 60 grep -qm1 'TERM' <(tail -n 1000 -f /var/log/manticore/searchd.log); then echo 'Searchd stopped.'; else echo 'Timeout or failed!'; cat /var/log/manticore/searchd.log; fi
```
I'm not sure `grep TERM` is enough to test that the searchd stopped successfully. There should be smth like 'Shutdown complete'

PavelShilin89 commented:
@sanikolaev All comments have been corrected, please approve the merge with master

sanikolaev commented on test/clt-tests/core/test-unfreeze-defrosts.rec:
```diff
@@ -62,3 +62,18 @@ sleep 5; mysql -h0 -P9306 -e "select count(*) from orders\G"
+| /var/lib/manticore/t/t.settings | /var/lib/manticore/t/t.settings |
++---------------------------------+---------------------------------+
+––– input –––
+searchd --stop > /dev/null 2>&1; if timeout 60 grep -qm1 'shutting down' <(tail -n 1000 -f /var/log/manticore/searchd.log); then echo 'Searchd stopped.'; else echo 'Timeout or failed!'; cat /var/log/manticore/searchd.log; fi
```
Using `shutting down` doesn't seem like the best way to check if Manticore has stopped. The correct term to search for is `shutdown complete`.

PavelShilin89 commented:
@sanikolaev I found the problem and fixed it, it was that `sleep 5; mysql -h0 -P9306 -e ‘unfreeze orders;’` locally runs slower than in CI, and further in the test when `searchd --stopwait` `unfreeze` was still running. In CI `unfreeze` ran faster by the time `searchd --stopwait` it had already completed, so I had no errors. To improve the test I added `timeout 60 bash -c ‘while [[ $(mysql -h0 -P9306 -e “SELECT COUNT(*) FROM orders;” | grep -o “[0-9]*”) -ne 10000 ]]]; do sleep 1; done && echo “Data processing completed.”’` to avoid using `sleep`. In that case everything works identically and correctly. 
I request approval to merge with the master.
